File: summary/elixir/4_bubble_sort.exs
defmodule BubbleSort do
  def sort(list) when is_list(list) do
    sort(list, length(list))
  end

  defp sort(list, 0), do: list
  defp sort(list, n) do
    list |> bubble(n) |> sort(n - 1)
  end

  defp bubble(list, 0), do: list
  defp bubble([], _), do: []
  defp bubble([a, b | t], n) when a > b do
    [b, a | bubble(t, n - 1)]
  end
  defp bubble([h | t], n), do: [h | bubble(t, n - 1)]
end

# 使用例
unsorted_list = [5, 3, 8, 4, 2, 1, 6, 7]
sorted_list = BubbleSort.sort(unsorted_list)
IO.inspect(sorted_list)

File: summary/dart/4_bubble_sort.dart
List<int> bubbleSort(List<int> list) {
  int n = list.length;
  int i, step;
  for (step = 0; step < n; step++) {
    for (i = 0; i < n - step - 1; i++) {
      if (list[i] > list[i + 1]) {
        // Swap elements if they are in wrong order
        int temp = list[i];
        list[i] = list[i + 1];
        list[i + 1] = temp;
      }
    }
  }
  return list;
}

void main() {
  List<int> unsortedList = [5, 3, 8, 4, 2, 1, 6, 7];
  List<int> sortedList = bubbleSort(unsortedList);
  print(sortedList);
}
